File: client/src/main/java/com/flipkart/grayskull/GrayskullClientImpl.java
package com.flipkart.grayskull;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;
import com.flipkart.grayskull.auth.GrayskullAuthHeaderProvider;
import com.flipkart.grayskull.constants.MDCKeys;
import com.flipkart.grayskull.metrics.MetricsPublisher;
import com.flipkart.grayskull.models.GrayskullClientConfiguration;
import com.flipkart.grayskull.models.response.HttpResponse;
import com.flipkart.grayskull.models.SecretValue;
import com.flipkart.grayskull.models.exceptions.GrayskullException;
import com.flipkart.grayskull.hooks.NoOpRefreshHandlerRef;
import com.flipkart.grayskull.hooks.RefreshHandlerRef;
import com.flipkart.grayskull.hooks.SecretRefreshHook;
import com.flipkart.grayskull.models.response.Response;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

/**
 * Implementation of the Grayskull client.
 * <p>
 * This class provides the main functionality for interacting with the Grayskull
 * secret management service, including fetching secrets and managing refresh hooks.
 * </p>
 */
public final class GrayskullClientImpl implements GrayskullClient {
    private static final Logger log = LoggerFactory.getLogger(GrayskullClientImpl.class);
    private static final TypeReference<Response<SecretValue>> SECRET_VALUE_TYPE_REFERENCE = 
            new TypeReference<Response<SecretValue>>() {};
    
    private final String baseUrl;
    private final GrayskullAuthHeaderProvider authHeaderProvider;
    private final GrayskullClientConfiguration grayskullClientConfiguration;
    private final GrayskullHttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final MetricsPublisher metricsPublisher;

    /**
     * Creates a new Grayskull client implementation.
     *
     * @param authHeaderProvider provider for authentication headers (must not be null)
     * @param grayskullClientConfiguration configuration properties (must not be null)
     * @throws IllegalArgumentException if authHeaderProvider or grayskullClientConfiguration is null
     */
    public GrayskullClientImpl(GrayskullAuthHeaderProvider authHeaderProvider, GrayskullClientConfiguration grayskullClientConfiguration) {
        
        if (authHeaderProvider == null) {
            throw new IllegalArgumentException("authHeaderProvider cannot be null");
        }
        if (grayskullClientConfiguration == null) {
            throw new IllegalArgumentException("grayskullClientConfiguration cannot be null");
        }
        
        this.baseUrl = grayskullClientConfiguration.getHost();
        this.authHeaderProvider = authHeaderProvider;
        this.grayskullClientConfiguration = grayskullClientConfiguration;
        this.httpClient = new GrayskullHttpClient(authHeaderProvider, grayskullClientConfiguration);
        
        this.objectMapper = new ObjectMapper();
        this.objectMapper.registerModule(new ParameterNamesModule());
        
        // Initialize metrics publisher if enabled
        this.metricsPublisher = grayskullClientConfiguration.isMetricsEnabled() ? new MetricsPublisher() : null;
    }
    
    /**
     * Retrieves a secret from the Grayskull server.
     * <p>
     * The secretRef should be in the format: "projectId:secretName"
     * For example: "my-project:database-password"
     * </p>
     *
     * @param secretRef the secret reference in format "projectId:secretName"
     * @return the secret value
     * @throws IllegalArgumentException if secretRef format is invalid
     * @throws RuntimeException if the secret cannot be retrieved
     */
    @Override
    public SecretValue getSecret(String secretRef) {
        String requestId = generateRequestId();
        MDC.put(MDCKeys.GRAYSKULL_REQUEST_ID, requestId);

        long startTime = System.nanoTime();
        
        int statusCode = 0;

        try {
            if (secretRef == null || secretRef.isEmpty()) {
                throw new IllegalArgumentException("secretRef cannot be null or empty");
            }

            // Parse secretRef format: "projectId:secretName"
            String[] parts = secretRef.split(":", 2);
            if (parts.length != 2) {
                throw new IllegalArgumentException(
                        "Invalid secretRef format. Expected 'projectId:secretName', got: " + secretRef);
            }

            String projectId = parts[0];
            String secretName = parts[1];

            if (projectId.isEmpty() || secretName.isEmpty()) {
                throw new IllegalArgumentException(
                        "projectId and secretName cannot be empty in secretRef: " + secretRef);
            }

            // Put context in MDC for automatic inclusion in all log statements
            MDC.put(MDCKeys.PROJECT_ID, projectId);
            MDC.put(MDCKeys.SECRET_NAME, secretName);

            log.debug("[RequestId:{}] Fetching secret", requestId);

            // URL encode the path parameters to handle special characters (spaces, slashes, etc.)
            String encodedProjectId = urlEncode(projectId);
            String encodedSecretName = urlEncode(secretName);
            String url = baseUrl + String.format("/v1/projects/%s/secrets/%s/data", encodedProjectId, encodedSecretName);
            
            // Fetch the secret with automatic retry logic
            HttpResponse httpResponse = httpClient.doGetWithRetry(url);
            statusCode = httpResponse.getStatusCode();
            
            Response<SecretValue> response = objectMapper.readValue(httpResponse.getBody(), SECRET_VALUE_TYPE_REFERENCE);
            SecretValue secretValue = response.getData();

            if (secretValue == null) {
                throw new GrayskullException(500, "No data in response");
            }
            
            return secretValue;
            
        } catch (JsonProcessingException e) {
            // JSON parsing errors are not retryable - they indicate a permanent problem
            throw new GrayskullException("Failed to parse response: ", e);
        } catch (GrayskullException e) {
            statusCode = e.getStatusCode();
            throw e;
        } finally {
            if (metricsPublisher != null) {
                long duration = System.nanoTime() - startTime;
                long durationMs = TimeUnit.NANOSECONDS.toMillis(duration);
                
                metricsPublisher.recordRequest("getSecret." + secretRef, statusCode, durationMs);
            }
            
            // Clean up MDC context
            MDC.remove(MDCKeys.PROJECT_ID);
            MDC.remove(MDCKeys.SECRET_NAME);
            MDC.remove(MDCKeys.GRAYSKULL_REQUEST_ID);
        }
    }

    /**
     * Registers a refresh hook for a secret.
     * <p>
     * Note: This is a placeholder implementation. The hook will be registered but
     * will not be invoked until server-side long-polling support is implemented.
     * This allows applications to include hook registration code now, and it will
     * automatically work when upgrading to a future SDK version with full support.
     * </p>
     *
     * @param secretRef the secret reference to monitor
     * @param hook the hook to invoke when the secret is refreshed
     * @return a handle that can be used to check status or unregister the hook
     */
    @Override
    public RefreshHandlerRef registerRefreshHook(String secretRef, SecretRefreshHook hook) {
        String requestId = generateRequestId();
        MDC.put(MDCKeys.GRAYSKULL_REQUEST_ID, requestId);

        if (secretRef == null || secretRef.isEmpty()) {
            throw new IllegalArgumentException("secretRef cannot be null or empty");
        }
        if (hook == null) {
            throw new IllegalArgumentException("hook cannot be null");
        }
        
        log.debug("[RequestId:{}] Registering refresh hook (placeholder implementation)", requestId);
        
        // TODO: Implement actual hook invocation when server-side events support is added
        return NoOpRefreshHandlerRef.INSTANCE;
    }

    /**
     * Closes the client and releases resources.
     * <p>
     * This method should be called when the client is no longer needed to properly
     * clean up HTTP connections and other resources.
     * </p>
     */
    @Override
    public void close() {
        log.info("Closing Grayskull client");
        if (httpClient != null) {
            httpClient.close();
        }
    }

    private static String urlEncode(String value) {
        try {
            return URLEncoder.encode(value, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            throw new GrayskullException(500, "Failed to URL encode value: " + value, e);
        }
    }

    private String generateRequestId() {
        return UUID.randomUUID().toString();
    }
}

File: clients/go/client_impl/grayskullClientImpl.go
package client_impl

import (
	"context"
	"errors"
	"fmt"
	"github.com/grayskull/client"
	"github.com/grayskull/client/hooks"
	"github.com/grayskull/client_impl/auth"
	"github.com/grayskull/client_impl/metrics"
	"log/slog"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/google/uuid"

	CLientModels "github.com/grayskull/client/models"
	"github.com/grayskull/client_impl/constants"
	ClientHooks "github.com/grayskull/client_impl/hooks"
	"github.com/grayskull/client_impl/models"
	"github.com/grayskull/client_impl/models/response"
)

// GrayskullClientImpl implements the Grayskull client interface
type GrayskullClientImpl struct {
	baseURL             string
	authHeaderProvider  auth.GrayskullAuthHeaderProvider
	clientConfiguration *models.GrayskullClientConfiguration
	httpClient          *GrayskullHTTPClient
	logger              *slog.Logger
	metrics             *metrics.Metrics
	metricsServer       *metrics.Server
}

// NewGrayskullClient creates a new instance of a Grayskull client
func NewGrayskullClient(authHeaderProvider auth.GrayskullAuthHeaderProvider, clientConfig *models.GrayskullClientConfiguration) (client.Client, error) {
	if authHeaderProvider == nil {
		return nil, errors.New("authHeaderProvider cannot be nil")
	}

	if clientConfig == nil {
		return nil, errors.New("clientConfig cannot be nil")
	}

	// Initialize logger with default JSON handler
	logger := slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{
		Level: slog.LevelInfo,
	}))

	// Initialize metrics
	metricsInstance := metrics.NewMetrics("grayskull_client")

	// Start metrics server if enabled
	var metricsServer *metrics.Server
	if clientConfig.Metrics.Enabled && clientConfig.Metrics.Server != nil {
		metricsServer = metrics.NewServer(clientConfig.Metrics.Server.Address)
		if err := metricsServer.Start(); err != nil {
			logger.Error("Failed to start metrics server", "error", err)
		} else {
			logger.Info("Metrics server started", "address", metricsServer.Addr())
		}
	}

	// Ensure base URL ends with a slash
	baseURL := clientConfig.Host
	if !strings.HasSuffix(baseURL, "/") {
		baseURL += "/"
	}

	// Initialize HTTP client
	httpClient, err := NewGrayskullHTTPClient(authHeaderProvider, clientConfig, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create HTTP client: %w", err)
	}

	client := &GrayskullClientImpl{
		baseURL:             baseURL,
		authHeaderProvider:  authHeaderProvider,
		clientConfiguration: clientConfig,
		httpClient:          httpClient,
		logger:              logger,
		metrics:             metricsInstance,
		metricsServer:       metricsServer,
	}

	return client, nil
}

// GetSecret retrieves a secret from the Grayskull server.
// The secretRef should be in the format: "projectId:secretName"
// For example: "my-project:database-password"
func (c *GrayskullClientImpl) GetSecret(ctx context.Context, secretRef string) (secretValue *CLientModels.SecretValue, operationErr error) {
	startTime := time.Now()
	requestID := uuid.NewString()

	// Add request ID to context before starting the operation
	ctx = context.WithValue(ctx, constants.GrayskullRequestID, requestID)

	// Defer function to record metrics
	defer func() {
		duration := time.Since(startTime).Seconds()
		status := "success"
		if operationErr != nil {
			status = "error"
		} else if ctx.Err() != nil {
			status = "canceled"
		}
		c.metrics.ObserveSecretOperationDuration("get_secret", status, duration)
		c.metrics.IncSecretOperationTotal("get_secret", status)
	}()

	// Set up logger with context
	logger := c.logger.With(
		constants.GrayskullRequestID, requestID,
		constants.SecretName, secretRef,
	)

	defer func() {
		duration := time.Since(startTime)
		logger.Info("getSecret completed",
			"duration_ms", duration.Milliseconds(),
		)
	}()

	if secretRef == "" {
		return nil, errors.New("secretRef cannot be empty")
	}

	// Parse secretRef format: "projectId:secretName"
	parts := strings.SplitN(secretRef, ":", 2)
	if len(parts) != 2 {
		return nil, fmt.Errorf("invalid secretRef format. Expected 'projectId:secretName', got: %s", secretRef)
	}

	projectID := parts[0]
	secretName := parts[1]

	if projectID == "" || secretName == "" {
		return nil, fmt.Errorf("projectId and secretName cannot be empty in secretRef: %s", secretRef)
	}

	// Add project and secret to logger context
	logger = logger.With(
		constants.ProjectID, projectID,
		constants.SecretName, secretName,
	)

	logger.Debug("fetching secret")

	// URL encode the path parameters
	encodedProjectID := url.PathEscape(projectID)
	encodedSecretName := url.PathEscape(secretName)
	requestURL := fmt.Sprintf("%sv1/projects/%s/secrets/%s/data", c.baseURL, encodedProjectID, encodedSecretName)

	// Fetch the secret with automatic retry logic
	httpResponse, err := c.httpClient.DoGetWithRetry(ctx, requestURL)
	if err != nil {
		logger.Error("failed to fetch secret", "error", err)
		return nil, fmt.Errorf("failed to fetch secret: %w", err)
	}

	secretResponse, err := response.UnmarshalResponse[CLientModels.SecretValue]([]byte(httpResponse.Body))
	if err != nil {
		logger.Error("failed to parse secret response", "error", err)
		return nil, fmt.Errorf("failed to parse secret response: %w", err)
	}

	if secretResponse == nil {
		return nil, errors.New("empty response from server")
	}

	// Get the actual secret data from the response
	secretData := secretResponse.Data

	// Check if the SecretValue is the zero value
	if secretData == (CLientModels.SecretValue{}) {
		return nil, errors.New("empty secret data in response")
	}

	return &secretData, nil
}

// RegisterRefreshHook registers a refresh hook for a secret.
// Note: This is a placeholder implementation. The hook will be registered but
// will not be invoked until server-side long-polling support is implemented.
func (c *GrayskullClientImpl) RegisterRefreshHook(ctx context.Context, secretRef string, hook hooks.SecretRefreshHook) (hooks.RefreshHandlerRef, error) {
	startTime := time.Now()
	status := "success"

	// Defer function to record metrics
	defer func() {
		duration := time.Since(startTime).Seconds()
		c.metrics.ObserveSecretOperationDuration("register_refresh_hook", status, duration)
		c.metrics.IncSecretOperationTotal("register_refresh_hook", status)
	}()

	// Validate inputs
	if secretRef == "" {
		status = "error"
		return nil, fmt.Errorf("secretRef cannot be empty")
	}
	if hook == nil {
		status = "error"
		return nil, fmt.Errorf("hook cannot be nil")
	}

	// Create a new no-op refresh handler ref
	handlerRef := &ClientHooks.NoOpRefreshHandlerRef{}

	// Log that a refresh hook was registered (for debugging)
	c.logger.DebugContext(ctx, "refresh hook registered",
		"secretRef", secretRef,
	)

	return handlerRef, nil
}

// Close releases resources used by the Grayskull client.
// This method should be called when the client is no longer needed to properly
// clean up HTTP connections and other resources.
func (c *GrayskullClientImpl) Close() error {
	var errs []error

	// Close the metrics server if it was started
	if c.metricsServer != nil {
		c.logger.Info("shutting down metrics server")
		if err := c.metricsServer.Stop(context.Background()); err != nil {
			errMsg := fmt.Sprintf("error stopping metrics server: %v", err)
			errs = append(errs, errors.New(errMsg))
			c.logger.Error("error stopping metrics server", "error", err)
		} else {
			c.logger.Info("metrics server stopped successfully")
		}
	}

	// Close the HTTP client
	if c.httpClient != nil {
		c.logger.Debug("closing HTTP client")
		if err := c.httpClient.Close(); err != nil {
			errMsg := fmt.Sprintf("error closing HTTP client: %v", err)
			errs = append(errs, errors.New(errMsg))
			c.logger.Error("error closing HTTP client", "error", err)
		} else {
			c.logger.Debug("HTTP client closed successfully")
		}
	}

	// Log final metrics summary if metrics were enabled
	if c.metrics != nil && c.clientConfiguration.Metrics.Enabled && c.clientConfiguration.Metrics.Server != nil {
		c.logger.Info("metrics collection completed",
			"metrics_endpoint", fmt.Sprintf("http://%s/metrics", c.clientConfiguration.Metrics.Server.Address),
		)
	}

	// Return combined errors if any occurred
	if len(errs) > 0 {
		var errMsg string
		for _, e := range errs {
			errMsg += e.Error() + "; "
		}
		return fmt.Errorf("errors during client shutdown: %s", errMsg)
	}

	return nil
}
